File: android/src/main/kotlin/com/example/flutter_get_device_battery_percent/FlutterGetDeviceBatteryPercentPlugin.kt
package com.example.flutter_get_device_battery_percent

import androidx.annotation.NonNull
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import android.content.Context
import android.content.ContextWrapper
import android.content.Intent
import android.content.IntentFilter
import android.os.BatteryManager
import android.os.Build.VERSION
import android.os.Build.VERSION_CODES
import java.text.SimpleDateFormat
import java.util.*

class FlutterGetDeviceBatteryPercentPlugin: FlutterPlugin, MethodCallHandler {
  private lateinit var channel: MethodChannel

  override fun onAttachedToEngine(@NonNull flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
    channel = MethodChannel(flutterPluginBinding.binaryMessenger, "flutter_get_device_battery_percent")
    channel.setMethodCallHandler(this)
  }

  override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: Result) {
    when (call.method) {
      "getBatteryLevel" -> {
        val batteryLevel = getBatteryLevel()
        if (batteryLevel != -1) {
          result.success(batteryLevel)
        } else {
          result.error("UNAVAILABLE", "Battery level not available.", null)
        }
      }
      "formattedDateTime" -> {
        val inputDateTime = call.argument<String>("dateTime")
        if (inputDateTime != null) {
          val formattedDate = formatDate(inputDateTime)
          result.success(formattedDate)
        } else {
          result.error("INVALID_ARGUMENT", "DateTime argument is missing.", null)
        }
      }
      else -> result.notImplemented()
    }
  }

  override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
    channel.setMethodCallHandler(null)
  }

  private fun getBatteryLevel(): Int {
    val batteryLevel: Int
    if (VERSION.SDK_INT >= VERSION_CODES.LOLLIPOP) {
      val batteryManager = getSystemService(Context.BATTERY_SERVICE) as BatteryManager
      batteryLevel = batteryManager.getIntProperty(BatteryManager.BATTERY_PROPERTY_CAPACITY)
    } else {
      val intent = ContextWrapper(applicationContext).registerReceiver(null, IntentFilter(Intent.ACTION_BATTERY_CHANGED))
      batteryLevel = intent!!.getIntExtra(BatteryManager.EXTRA_LEVEL, -1) * 100 / intent.getIntExtra(BatteryManager.EXTRA_SCALE, -1)
    }
    return batteryLevel
  }

  private fun formatDate(inputDateTime: String): String {
    val inputFormat = SimpleDateFormat("yyyy-MM-dd HH:mm:ss", Locale.getDefault())
    val outputFormat = SimpleDateFormat("yyyy:MM:dd", Locale.getDefault())
    return try {
      val date = inputFormat.parse(inputDateTime)
      outputFormat.format(date)
    } catch (e: Exception) {
      e.printStackTrace()
      "Error formatting date"
    }
  }
}

File: ios/Classes/FlutterGetDeviceBatteryPercentPlugin.swift
import Flutter
import UIKit

public class FlutterGetDeviceBatteryPercentPlugin: NSObject, FlutterPlugin {
  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(name: "flutter_get_device_battery_percent", binaryMessenger: registrar.messenger())
    let instance = FlutterGetDeviceBatteryPercentPlugin()
    registrar.addMethodCallDelegate(instance, channel: channel)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "getBatteryLevel":
      receiveBatteryLevel(result: result)
    case "formattedDateTime":
      if let arguments = call.arguments as? [String: Any],
         let inputDateTime = arguments["dateTime"] as? String {
        let formattedDate = formatDate(inputDateTime: inputDateTime)
        result(formattedDate)
      } else {
        result(FlutterError(code: "INVALID_ARGUMENT", message: "DateTime argument is missing or invalid.", details: nil))
      }
    default:
      result(FlutterMethodNotImplemented)
    }
  }

  private func receiveBatteryLevel(result: FlutterResult) {
    #if targetEnvironment(simulator)
    result(100) // Return a dummy value for simulator
    #else
    let device = UIDevice.current
    device.isBatteryMonitoringEnabled = true
    if device.batteryState == UIDevice.BatteryState.unknown {
      result(FlutterError(code: "UNAVAILABLE", message: "Battery level not available.", details: nil))
    } else {
      result(Int(device.batteryLevel * 100))
    }
    #endif
  }

  private func formatDate(inputDateTime: String) -> String {
    let inputFormat = DateFormatter()
    inputFormat.dateFormat = "yyyy-MM-dd HH:mm:ss"

    let outputFormat = DateFormatter()
    outputFormat.dateFormat = "yyyy:MM:dd"

    if let date = inputFormat.date(from: inputDateTime) {
      return outputFormat.string(from: date)
    } else {
      return "Error formatting date"
    }
  }
}
